### PR TITLE
feat: Support custom state IDs in `meltano run` with `--state-id-suffix`

### LINKED ISSUE #6164: Feature proposal: dynamic state IDs based on an input variable
This was called out as "nice to have" but not delivered in the initial iteration of auto-generated state IDs for `meltano run`. 

For scenarios where the 'environment' actually builds to multiple deploy targets, such as one per customer or one per CI/CD build number or one per CI/CD branch name, it would be great to have a configurable and dynamic `state ID suffix` that could take one or more inputs, and use those inputs to ensure a unique set of STATE ID refs when running via `meltano run`.

Today state IDs are generated like:

```
${environment_name}:${tap-name}-to-${target-name}
```

But we could add some per-environment setting such as `state_id_suffix` or `deploy_target_key` which would concatenate with the base state ID calc:

```
${environment_name}:${tap-name}-to-${target-name}:${suffix_str}
```

Taking the branch names example, and an environment name called `cicd`, this could look like the following:

```
cicd:tap-github-to-target-snowflake:main
cicd:tap-github-to-target-snowflake:develop
cicd:tap-github-to-target-snowflake:staging
cicd:tap-github-to-target-snowflake:feat/my-cool-feature
cicd:tap-github-to-target-snowflake:fix/my-super-important-fix
```

Where each of these variations are basically populated with some input like `CI_BRANCH_REF`.

### LINKED ISSUE #6370: Add ability to provide custom `--state-id-suffix` to `meltano run`
One thing that was cut late in development when adding incremental job support to `meltano run` was the ability to add a custom prefix to the auto generated job (now `state`) names: https://gitlab.com/meltano/meltano/-/merge_requests/2544#note_855957749

> I think we've agreed on this already but just want to reiterate given that we've just logged [ADR: Refactor/rename "job_id" as "state-id" (#3290 - closed)](https://github.com/meltano/meltano/-/issues/3290), it is probably best to try to avoid user config/override for the state_id_prefix or job_id_prefix. Just for the reason that if we allow them to configure the naming convention, and then we update what it is called, that will create a breaking change which I think it's best to avoid.

Now that the `job` -> `state` transition work has basically wrapped up we could revisit adding this.

@tayloramurphy fwiw this is a pretty easy-peasy change on the dev side. If only configurable as a cli arg weights probably as low as a 2, but if we want to also support this as a per env meltano.yml option than most likely more realistically a weight of 4.

### PR DESCRIPTION
**Pull request by ReubenFrankel**
Resolves #6370, resolves #6164

### Checklist
- [x] `--state-id-suffix` option for `meltano run`
- [x] `state_id_suffix` option for environments in `meltano.yml`
- [x] Support expansion of environment variables for `state_id_suffix` value
- [x] Add/update tests
- [x] Update docs

### CODE REVIEW

**Comment by ReubenFrankel** [comment]
I would like to add an end-to-end test for `--state-id-suffix` in [`test_run.py`](https://github.com/meltano/meltano/blob/fba29109eb36cde61d652e17960a241ed5d8edca/tests/meltano/cli/test_run.py) to check a state ID is generated with the given suffix, but the existing tests are quite complex so not sure where to start. Any pointers?

**Comment by aaronsteers** [comment]
> I would like to add an end-to-end test for `--state-id-suffix` in [`test_run.py`](https://github.com/meltano/meltano/blob/fba29109eb36cde61d652e17960a241ed5d8edca/tests/meltano/cli/test_run.py) to check a state ID is generated with the given suffix, but the existing tests are quite complex so not sure where to start. Any pointers?

@pandemicsyn, @edgarrmondragon - Any ideas on the above?

@ReubenFrankel - I just noticed you'd kicked off this PR - very exciting! 🚀

**Comment by edgarrmondragon** [comment]
> > I would like to add an end-to-end test for `--state-id-suffix` in [`test_run.py`](https://github.com/meltano/meltano/blob/fba29109eb36cde61d652e17960a241ed5d8edca/tests/meltano/cli/test_run.py) to check a state ID is generated with the given suffix, but the existing tests are quite complex so not sure where to start. Any pointers?
> 
> @pandemicsyn, @edgarrmondragon - Any ideas on the above?
> 
> @ReubenFrankel - I just noticed you'd kicked off this PR - very exciting! 🚀

Probably copying and modifying the first bit of `TestCliRunScratchpadOne.test_run_basic_invocations`: https://github.com/meltano/meltano/blob/e4d698495e50ee9d0a81d4885dced8f79f519982/tests/meltano/cli/test_run.py#L295-L340

and using the `job_logging_service` to check that the right state ID was used to create the job log

https://github.com/meltano/meltano/blob/e4d698495e50ee9d0a81d4885dced8f79f519982/src/meltano/core/logging/job_logging_service.py#L55-L56

File: src/meltano/core/plugin/settings_service.py
Comment (ReubenFrankel) [inline src/meltano/core/plugin/settings_service.py:85]: Implementation related to #6770.

**Comment by ReubenFrankel** [comment]
Unrelated test failure: https://github.com/meltano/meltano/actions/runs/3087082213/jobs/4992077395#step:11:3279

File: src/meltano/core/plugin/settings_service.py
Comment (aaronsteers) [reply src/meltano/core/plugin/settings_service.py:85]: @kgpayne - You and @cjohnhanson are our experts on config and env var parsing as of now. 😄 Can you take a look at the above and advise?

**Comment by pandemicsyn** [comment]
> Unrelated test failure: https://github.com/meltano/meltano/actions/runs/3087082213/jobs/4992077395#step:11:3279

Looks like it cleared on subsequent updates.  Based on the conn resets, i'm guessing it was just transient and the host/vm had a networking issues.

File: tests/meltano/core/block/test_extract_load.py
Comment (pandemicsyn) [inline tests/meltano/core/block/test_extract_load.py:644]:
```suggestion
        )
        
        assert (
            generate_state_id(project, None, block1, block2)
            == "test:block1-to-block2"
        )

```

Worth testing no suffix just incase ?

File: tests/meltano/cli/test_run.py
Comment (pandemicsyn) [inline tests/meltano/cli/test_run.py:492]: nice 🤘 IMHO this should cover us for the change. 

@aaronsteers would be nice to do a integration test to show case/cover some of the state manipulation and behaviors that we support now (pop'd https://github.com/meltano/meltano/issues/6773 for that for us to follow up on).

File: src/meltano/core/state_service.py
Comment (pandemicsyn) [reply src/meltano/core/state_service.py:19]: Guessing pulling STATE_ID_COMPONENT_DELIMETER and `meltano.core.block.extract_load.generate_state_id` into meltano.core.utils or the like probably doesn't help ?

**Comment by pandemicsyn** [comment]
@ReubenFrankel looking great! IMHO you're safe to include the doc update whenever you're ready.

File: tests/meltano/core/block/test_extract_load.py
Comment (ReubenFrankel) [reply tests/meltano/core/block/test_extract_load.py:643]: Oops, that was my original intention... :sweat_smile: https://github.com/meltano/meltano/pull/6758/commits/fc2af322ff9d9aa1ef37b906af609b265ebf8b3d

File: src/meltano/core/state_service.py
Comment (ReubenFrankel) [reply src/meltano/core/state_service.py:19]: I'll have a look to see if that works. Would that be preferable?

File: docs/src/_concepts/environments.md
Comment (tayloramurphy) [inline docs/src/_concepts/environments.md:73]:
```suggestion
## State ID Suffix
```

File: docs/src/_concepts/environments.md
Comment (tayloramurphy) [inline docs/src/_concepts/environments.md:75]:
```suggestion
Environments can also define a `state_id_suffix` - a custom suffix used in the generation of a state ID for each extractor/loader pair passed to [`meltano run`]((/reference/command-line-interface#run)).
The suffix is appended to the generated ID with a colon prefix: `:<state_id_suffix>`. 

The full ID when a suffix is present is `<environment_name>:<tap_name>-to-<target_name>:<state_id_suffix>`.
```

File: docs/src/_concepts/environments.md
Comment (tayloramurphy) [inline docs/src/_concepts/environments.md:79]: This section goes with the "The env mapping" section, not with the state id suffix. It needs to be moved back up. Check the preview https://deploy-preview-6758--meltano.netlify.app/concepts/environments#state_id_suffix

**Review by tayloramurphy — CHANGES_REQUESTED** [review]

File: src/meltano/core/state_service.py
Comment (pandemicsyn) [reply src/meltano/core/state_service.py:19]: Thinking about , I bet you'd still end up with a circular import. The annoy bit is that generate_state_id wants a consumer/producer IOBlock , but *only* uses the string_id's of those. One way to clean this up then would be to update `generate_state_id` to just expect a string for consumer/producer and just have callers start passing the string_id directly rather than passing the whole IOBlock.

That's a bit more work/refactoring though - I think you could keep the pr as-is and then we can log a follow up issue to clean up/refactor.

**Review by pandemicsyn — APPROVED** [review]
lgtm - pending the doc changes @tayloramurphy flagged

File: docs/src/_concepts/environments.md
Comment (ReubenFrankel) [reply docs/src/_concepts/environments.md:79]: My bad, fixing now :+1:

File: docs/src/_concepts/environments.md
Comment (ReubenFrankel) [reply docs/src/_concepts/environments.md]: See https://github.com/meltano/meltano/pull/6758/commits/fdca026b0c5376d23e7d147e19693e3b13003a3d

File: docs/src/_concepts/environments.md
Comment (ReubenFrankel) [reply docs/src/_concepts/environments.md]: See See https://github.com/meltano/meltano/pull/6758/commits/fdca026b0c5376d23e7d147e19693e3b13003a3d

File: docs/src/_concepts/environments.md
Comment (ReubenFrankel) [reply docs/src/_concepts/environments.md]: See https://github.com/meltano/meltano/pull/6758/commits/fdca026b0c5376d23e7d147e19693e3b13003a3d

File: src/meltano/core/state_service.py
Comment (ReubenFrankel) [reply src/meltano/core/state_service.py:19]: Sure :+1:

**Comment by ReubenFrankel** [comment]
@pandemicsyn I've resolved all conversations I think are complete, just waiting on the others.

**Review by tayloramurphy — APPROVED** [review]
Docs approval

File: src/meltano/core/plugin/settings_service.py
Comment (kgpayne) [reply src/meltano/core/plugin/settings_service.py:85]: Thanks for this @ReubenFrankel 🙌 I'm not surprised that env var expansion was a hurdle - we have been talking for a while about centralising env var handling onto its own service (#6253). Your approach is inline with the current state of env var expansion 👍 I was surprised that the `self.project.dotenv_env` was necessary, as I'd have expected this to be filled in already at the `project_settings_service` layer. I see it comes in stead from `get_with_metadata` when accessing 'normal' settings, so is needed here. Will put more thoughts along that line in its own comment...

**Comment by kgpayne** [comment]
Following this comment on [env var expansion](https://github.com/meltano/meltano/pull/6758#discussion_r977547766), it strikes me that manual expansion is needed in this case because `state_id_suffix` is implemented as a new attribute of the `Environment`, and so bypasses the `get`/`get_with_metadata` method on the `SettingsService` (and descendants), which handles env var expansion for 'normal' settings. I wouldn't block on it for now, but in future we might choose to take the same approach as with plugins; treating extra `Environment` attributes as [extras](https://docs.meltano.com/guide/configuration#plugin-extras), accessible via the regular `get_with_metadata` and `set_with_metadata` calls (which handle expansion). Apart from simplifying expansion, this would also make it much easier to add new "Environment setting extras" like this in future, accessible via `project_setting_service.get_with_metadata("_extra.state_id_suffix")` 🙂

I think this would be as simple as updating the `MeltanoEnvStoreManager` to return any `Environment` attrs other than `['name', 'config' 'env']` as `_extras` in its `flat_config` property, but I haven't looked too closely. Configuration and overrides would happen in the usual settings hierarchy (env vars, yml etc.), and the cli can be kept in addition (if required) by passing the given value through as `config_override` to the `ProjectSettingsService` constructor.

Interested in your thoughts @edgarrmondragon, and if you think is worth capturing this in a new issue 👍 FYI @aaronsteers

**Review by kgpayne — CHANGES_REQUESTED** [review]
Ah, actually I think these two need swapping, in line with how we interpolate in `get_with_metadata` [here](https://github.com/meltano/meltano/blob/main/src/meltano/core/settings_service.py#L333) 😅

File: src/meltano/core/plugin/settings_service.py
Comment (kgpayne) [inline src/meltano/core/plugin/settings_service.py:83]:
```suggestion
                    {
                        **self.project.dotenv_env,
                        **self.env_override,
                    },
```

**Review by kgpayne — DISMISSED** [review]

File: tests/meltano/cli/test_run.py
Comment (aaronsteers) [reply tests/meltano/cli/test_run.py:492]: Thanks, @pandemicsyn  👍 

Much appreciated.

**Review by aaronsteers — APPROVED** [review]
Looks great! Thanks, @ReubenFrankel for contributing.

**Pull request merged**
